File: Company 5 : Atlassian/Day 2/6. Find Beautiful Indices in the Given Array I/code.java
class Solution {
    public List<Integer> beautifulIndices(String s, String a, String b, int k) {
        // List to store the result
        List<Integer> result = new ArrayList<>();
        // Find the first occurrences of strings 'a' and 'b' in 's'
        int a_pos = s.indexOf(a); 
        int b_pos = s.indexOf(b);
        // Loop until we run out of occurrences for either 'a' or 'b'
        while (a_pos >= 0 && b_pos >= 0) {
            // If the distance between positions is within the allowed range (<= k)
            if (Math.abs(a_pos - b_pos) <= k) {
                result.add(a_pos); // Add the index of 'a' to the result
                a_pos = s.indexOf(a, a_pos + 1); // Move to the next occurrence of 'a'
            } 
            // If the position of 'b' is less than 'a', move to the next occurrence of 'b'
            else if (b_pos < a_pos) {
                b_pos = s.indexOf(b, b_pos + 1);
            } 
            // Otherwise, move to the next occurrence of 'a' and reset 'b' position
            else {
                a_pos = s.indexOf(a, a_pos + 1);
                b_pos = s.indexOf(b);
            }
        }
        return result; // Return the list of beautiful indices
    }
}

File: Company 5 : Atlassian/Day 2/6. Find Beautiful Indices in the Given Array I/code.cpp
class Solution {
public:
    vector<int> findOccurrences(const string& text, const string& pattern) {
        vector<int> occurrences;
        int pos = text.find(pattern, 0);
        while (pos != string::npos) {
            occurrences.push_back(pos);
            pos = text.find(pattern, pos + 1);
        }
        return occurrences;
    }
    
    vector<int> beautifulIndices(string s, string a, string b, int k) {
        vector<int> indices_a = findOccurrences(s, a);
        vector<int> indices_b = findOccurrences(s, b);
        vector<int> result;
        int j = 0;
        for (int i : indices_a) {
            while (j < indices_b.size() && indices_b[j] < i - k) {
                j++;
            }
            if (j < indices_b.size() && indices_b[j] <= i + k) {
                result.push_back(i);
            }
        }

        return result;
    }
};